File: anna/portfolio.js
// anna/portfolio.js
// ─────────────────────────────────────────────────────────────────────────────
// Driver Portfolio (Stage 3) + compliance summary writing (Stage 4).
//
// buildPortfolio()  aggregates everything Anna has collected about a driver into
//                   one reviewable record, and assigns a recruiter.
// writeCompliance() compares pulled MVR/PSP/Clearinghouse data against the
//                   matched carrier's gates and produces an approve/reject flag
//                   with per-category reasons + a human-readable summary. Uses
//                   the deterministic gate engine for the verdict (auditable) and
//                   optionally Claude to phrase the narrative summary.
// ─────────────────────────────────────────────────────────────────────────────

import { compileSpec } from './spec.js';
import { evaluateGates, STATUS } from './matcher.js';
import { callClaude, MODELS, annaConfigured } from './claude.js';

let _seq = 0;
const id = (p) => `${p}_${Date.now().toString(36)}${(_seq++).toString(36)}`;

/**
 * Build a driver portfolio with Anna's RANKED carrier recommendations.
 *
 * Anna does NOT pick the carrier — she ranks the best fits and a human recruiter
 * selects one (see selectCarrier). So the portfolio starts with `carrier: null`
 * and a `recommendations` list, awaiting recruiter selection.
 *
 * @param {Object} p
 * @param {Object} p.driver          Normalized DriverProfile.
 * @param {Object} p.match           Full result from matchDriver ({ matches, top, summary }).
 * @param {Object} [p.documents]     { docType: { fields, confidence, warnings } }.
 * @param {string} [p.recruiter]     Pre-assigned recruiter, else round-robin.
 * @param {string[]} [p.recruiterPool]
 * @returns {Object} portfolio
 */
export function buildPortfolio({ driver, match, documents = {}, recruiter, recruiterPool = [] }) {
  const assigned = recruiter || pickRecruiter(recruiterPool);
  const docWarnings = Object.entries(documents).flatMap(([t, d]) => (d.warnings || []).map((w) => `${t}: ${w}`));
  const matches = (match && match.matches) || [];
  // Compact, ranked recommendation list for the recruiter to choose from.
  const recommendations = matches.map((m) => ({
    carrierId: m.carrierId,
    carrierName: m.carrierName,
    status: m.status,
    fitScore: m.fitScore,
    scoreBreakdown: m.scoreBreakdown || {},
    fitSummary: m.fitSummary || null,
    nearMiss: m.nearMiss || null,
    topReason: (m.reasons && m.reasons[0]) || null,
  }));
  return {
    id: id('portfolio'),
    createdAt: new Date().toISOString(),
    driver,
    recommendations,            // Anna's ranked best-fit list (advisory)
    suggestedTop: match?.top?.carrierId || null, // highlight only; NOT auto-selected
    carrier: null,              // set by the recruiter via selectCarrier()
    documents,
    documentWarnings: docWarnings,
    compliance: null,           // filled in by writeCompliance() after a carrier is picked
    review: {
      assignedRecruiter: assigned,
      task: assigned ? `${assigned}, please review this driver and pick the best-fit carrier.` : 'Unassigned — needs a recruiter.',
      status: 'awaiting_carrier', // awaiting_carrier | pending | approved | rejected
      carrierSelectedBy: null, carrierSelectedAt: null,
      decidedBy: null, decidedAt: null, decisionReason: null,
    },
  };
}

/**
 * Recruiter picks a carrier from the recommendations (the human-in-the-loop step).
 * Binds the portfolio to that carrier and moves it to 'pending' review.
 * @returns {Object} the updated portfolio (mutated in place).
 */
export function selectCarrier(portfolio, carrierId, by = 'Recruiter') {
  const rec = (portfolio.recommendations || []).find((r) => r.carrierId === carrierId);
  if (!rec) throw new Error('That carrier is not in this driver\'s recommendation list.');
  portfolio.carrier = { carrierId: rec.carrierId, carrierName: rec.carrierName, fitScore: rec.fitScore, status: rec.status, scoreBreakdown: rec.scoreBreakdown || {} };
  portfolio.review = {
    ...portfolio.review,
    status: portfolio.review.status === 'awaiting_carrier' ? 'pending' : portfolio.review.status,
    carrierSelectedBy: by,
    carrierSelectedAt: new Date().toISOString(),
  };
  return portfolio;
}

let _rr = 0;
function pickRecruiter(pool) {
  if (!pool.length) return null;
  return pool[_rr++ % pool.length];
}

/**
 * Run a compliance check: merge pulled records into the driver profile, evaluate
 * against the carrier's gates, and produce the approve/reject summary.
 *
 * @param {Object} p
 * @param {Object} p.carrier   Carrier record (with .requirements) or compiled spec.
 * @param {Object} p.driver    DriverProfile.
 * @param {Object} [p.records] { mvr, psp, clearinghouse } pulled from integrations,
 *                              shaped to merge into the driver profile.
 * @param {Object} [p.opts]    { apiKey, model, narrate }
 * @returns {Promise<Object>} compliance result
 */
export async function writeCompliance({ carrier, driver, records = {}, opts = {} }) {
  const spec = carrier.hardGates ? carrier : compileSpec(carrier);
  const merged = mergeRecords(driver, records);
  const { status, gateResults, failed, unknown } = evaluateGates(spec, merged);

  const flag = status === STATUS.ELIGIBLE ? 'approve' : status === STATUS.NEEDS_DATA ? 'review' : 'reject';
  const byCategory = groupByCategory(gateResults);
  let summary = buildSummary(flag, failed, unknown, spec.carrierName);

  // Optionally let Claude phrase a richer narrative; never let it change the verdict.
  if (opts.narrate && annaConfigured(opts.apiKey)) {
    try {
      const { text } = await callClaude({
        apiKey: opts.apiKey,
        model: opts.model || MODELS.smart,
        maxTokens: 400,
        system: 'You are Anna, an FMCSA driver-qualification compliance assistant. Write a concise, factual 2-3 sentence summary for a recruiter. Do not change the provided verdict. Cite specifics.',
        messages: [{ role: 'user', content: `Verdict: ${flag.toUpperCase()} for ${spec.carrierName}.\nGate results:\n${JSON.stringify(gateResults, null, 2)}\n\nWrite the summary.` }],
      });
      if (text.trim()) summary = text.trim();
    } catch { /* keep deterministic summary */ }
  }

  return {
    flag, // approve | reject | review
    status,
    carrierId: spec.carrierId,
    carrierName: spec.carrierName,
    categories: byCategory,
    reasons: [...failed, ...unknown].map((r) => r.reason),
    summary,
    checkedAt: new Date().toISOString(),
  };
}

// Merge pulled compliance records into the driver profile (records win).
// Exported so callers can persist the authoritative profile after a pull.
export function mergeRecords(driver, records) {
  return {
    ...driver,
    mvr: { ...driver.mvr, ...records.mvr },
    psp: { ...driver.psp, ...records.psp },
    // Clearinghouse: a positive/incomplete result is a hard disqualifier; expose
    // it as a DUI-equivalent so existing gates catch it (extendable later).
    ...(records.clearinghouse?.prohibited ? { mvr: { ...driver.mvr, ...records.mvr, dui: Math.max(driver.mvr?.dui || 0, 1) } } : {}),
  };
}

function groupByCategory(gateResults) {
  const map = {};
  for (const r of gateResults) {
    (map[r.category] ||= { pass: true, reasons: [] });
    if (r.status !== 'PASS') { map[r.category].pass = false; map[r.category].reasons.push(r.reason); }
  }
  return Object.entries(map).map(([key, v]) => ({ key, pass: v.pass, reason: v.reasons.join(' ') || `${key} meets requirements.` }));
}

function buildSummary(flag, failed, unknown, carrierName) {
  if (flag === 'approve') return `Driver meets all of ${carrierName}'s stated CDL, MVR, PSP, and insurance requirements and is recommended for approval.`;
  if (flag === 'reject') return `Driver does not meet ${carrierName}'s requirements: ${failed.map((f) => f.reason).join(' ')}`;
  return `Cannot finalize for ${carrierName} — missing data: ${unknown.map((u) => u.reason).join(' ')}`;
}

File: service/anna/portfolio.py
"""Portfolio build / carrier selection / compliance verdict — port of anna/portfolio.js."""
from __future__ import annotations
import time
from datetime import datetime, timezone
from .spec import compile_spec
from .matcher import evaluate_gates, STATUS

_seq = 0


def _now():
    return datetime.now(timezone.utc).isoformat()


def _id(prefix):
    global _seq
    _seq += 1
    return f"{prefix}_{int(time.time()*1000):x}{_seq:x}"


_rr = 0


def _pick_recruiter(pool):
    global _rr
    if not pool:
        return None
    r = pool[_rr % len(pool)]
    _rr += 1
    return r


def build_portfolio(driver, match, documents=None, recruiter=None, recruiter_pool=None):
    documents = documents or {}
    assigned = recruiter or _pick_recruiter(recruiter_pool or [])
    matches = (match or {}).get("matches") or []
    recommendations = [{
        "carrierId": m["carrierId"], "carrierName": m["carrierName"], "status": m["status"],
        "fitScore": m["fitScore"], "scoreBreakdown": m.get("scoreBreakdown") or {},
        "fitSummary": m.get("fitSummary"), "nearMiss": m.get("nearMiss"),
        "topReason": (m.get("reasons") or [None])[0],
    } for m in matches]
    doc_warnings = [f"{t}: {w}" for t, d in documents.items() for w in (d.get("warnings") or [])]
    return {
        "id": _id("portfolio"), "createdAt": _now(), "driver": driver,
        "recommendations": recommendations,
        "suggestedTop": ((match or {}).get("top") or {}).get("carrierId"),
        "carrier": None, "documents": documents, "documentWarnings": doc_warnings,
        "compliance": None,
        "review": {
            "assignedRecruiter": assigned,
            "task": f"{assigned}, please review this driver and pick the best-fit carrier." if assigned else "Unassigned — needs a recruiter.",
            "status": "awaiting_carrier",
            "carrierSelectedBy": None, "carrierSelectedAt": None,
            "decidedBy": None, "decidedAt": None, "decisionReason": None,
        },
    }


def select_carrier(portfolio, carrier_id, by="Recruiter"):
    rec = next((r for r in (portfolio.get("recommendations") or []) if r["carrierId"] == carrier_id), None)
    if not rec:
        raise ValueError("That carrier is not in this driver's recommendation list.")
    portfolio["carrier"] = {"carrierId": rec["carrierId"], "carrierName": rec["carrierName"],
                            "fitScore": rec["fitScore"], "status": rec["status"],
                            "scoreBreakdown": rec.get("scoreBreakdown") or {}}
    r = portfolio["review"]
    r["status"] = "pending" if r["status"] == "awaiting_carrier" else r["status"]
    r["carrierSelectedBy"] = by
    r["carrierSelectedAt"] = _now()
    return portfolio


def merge_records(driver, records):
    records = records or {}
    out = {**driver}
    out["mvr"] = {**(driver.get("mvr") or {}), **(records.get("mvr") or {})}
    out["psp"] = {**(driver.get("psp") or {}), **(records.get("psp") or {})}
    if (records.get("clearinghouse") or {}).get("prohibited"):
        out["mvr"] = {**out["mvr"], "dui": max((driver.get("mvr") or {}).get("dui") or 0, 1)}
    return out


def _group_by_category(gate_results):
    m = {}
    for r in gate_results:
        cat = m.setdefault(r["category"], {"pass": True, "reasons": []})
        if r["status"] != "PASS":
            cat["pass"] = False
            cat["reasons"].append(r["reason"])
    return [{"key": k, "pass": v["pass"], "reason": " ".join(v["reasons"]) or f"{k} meets requirements."}
            for k, v in m.items()]


def _build_summary(flag, failed, unknown, carrier_name):
    if flag == "approve":
        return f"Driver meets all of {carrier_name}'s stated requirements and is recommended for approval."
    if flag == "reject":
        return f"Driver does not meet {carrier_name}'s requirements: " + " ".join(f["reason"] for f in failed)
    return f"Cannot finalize for {carrier_name} — missing data: " + " ".join(u["reason"] for u in unknown)


async def write_compliance(carrier, driver, records=None, opts=None):
    records, opts = records or {}, opts or {}
    spec = carrier if carrier.get("hardGates") else compile_spec(carrier)
    merged = merge_records(driver, records)
    ev = evaluate_gates(spec, merged)
    flag = "approve" if ev["status"] == STATUS["ELIGIBLE"] else ("review" if ev["status"] == STATUS["NEEDS_DATA"] else "reject")
    summary = _build_summary(flag, ev["failed"], ev["unknown"], spec["carrierName"])
    if opts.get("narrate") and opts.get("apiKey"):
        try:
            from .llm import llm_complete
            res = await llm_complete(
                provider=opts.get("provider"), apiKey=opts.get("apiKey"), model=opts.get("model"), maxTokens=400,
                system=("You are Anna, an FMCSA driver-qualification compliance assistant. Write a concise, factual "
                        "2-3 sentence summary for a recruiter. Do not change the provided verdict. Cite specifics."),
                messages=[{"role": "user", "content":
                           f"Verdict: {flag.upper()} for {spec['carrierName']}.\nGate results:\n{ev['gateResults']}\n\nWrite the summary."}],
            )
            if (res.get("text") or "").strip():
                summary = res["text"].strip()
        except Exception:
            pass
    return {
        "flag": flag, "status": ev["status"], "carrierId": spec["carrierId"], "carrierName": spec["carrierName"],
        "categories": _group_by_category(ev["gateResults"]),
        "reasons": [r["reason"] for r in ev["failed"] + ev["unknown"]],
        "summary": summary, "checkedAt": _now(),
    }
